src/cs50/sql.py

```diff
@@ -286,11 +286,17 @@ def execute(self, sql, *args, **kwargs):
                 # Connect now
                 flask.g._connection = self._engine.connect()
 
-                # Disconnect later
-                @flask.current_app.teardown_appcontext
-                def shutdown_session(exception=None):
-                    if hasattr(flask.g, "_connection"):
-                        flask.g._connection.close()
+                # Disconnect later - but only once
+                teardown = True
+                for func in flask.current_app.teardown_appcontext_funcs:
+                    if func.__name__ == "shutdown_session":
+                        teardown = False
+
+                if teardown:
+                    @flask.current_app.teardown_appcontext
+                    def shutdown_session(exception=None):
+                        if hasattr(flask.g, "_connection"):
+                            flask.g._connection.close()
```

Comment:
I wonder if it would make more sense for the flag to be set on the `execute` function itself such that it could do something like:

```
def execute(...):
  ...
 
  # Ensure shutdown_session is just added once
  if not execute.teardown_appcontext_added:
    execute.teardown_appcontext_added = True

    # Register shutdown_session on app context teardown    
    @flask_current_app.teardown_appcontext
    def shutdown(...):
      ...

# teardown_appcontext is not added initially
execute.teardown_appcontext_added = False
```

instead of having to search through the list of `teardown_appcontext_funcs` repeatedly? Also, fwiw, I don't think function names uniquely identifies a function in this case (i.e., more than one `teardown_appcontext` handler could have the same name).

Reply:
Created an instance variable, `teardown_appcontext_added`, to the SQL class to do this.